lib/conn_parser.ex

```diff
@@ -1,62 +1,80 @@
+
+  Each connection sent to documenting in your tests is given to `ConnParser`.
+  Is expected that connection has been passed through the app `Endpoint` as a
+  finished request. The parser will extract all needed info from `Conn` and uses
+  app `Router` for additional information about the request.
+
+  The atribute `description` must be given at `document` macro call with the
+  option `:as`:
+
+      test "test name", %{conn: conn} do
+        ...
+
+        Xcribe.Helpers.Document.document(conn, as: "description here")
+
+        ...
+      end
+
+  If no description is given the current test description will be used.
+  """
+
   alias Xcribe.{Config, Request}
 
-  def execute(conn, description \\ "sample request") do
-    route = identify_route(conn)
-    path = format_path(route.path, Map.keys(conn.path_params))
-    namespaces = fetch_namespaces()
+  @doc """
+  Parse the given `Plug.Conn` and transform it to a `Xcribe.Request`. A
+  description can be provided.
+  """
+  def execute(conn, description \\ "") do
+    conn
+    |> identify_route()
+    |> parse_conn(conn, description)
+  end
+
+  defp parse_conn({:error, _} = error, _conn, _description), do: error
+
+  defp parse_conn(route, conn, description) do
+    path = format_path(route.route, Map.keys(conn.path_params))
 
     %Request{
       action: route |> router_options() |> Atom.to_string(),
       header_params: conn.req_headers,
-      controller: conn |> controller_module(),
+      controller: controller_module(route),
       description: description,
       params: conn.params,
       path: path,
       path_params: conn.path_params,
       query_params: conn.query_params,
       request_body: conn.body_params,
-      resource: resource_name(path, namespaces),
+      resource: resource_name(path, fetch_namespaces()),
       resource_group: resource_group(route),
       resp_body: conn.resp_body,
       resp_headers: conn.resp_headers,
       status_code: conn.status,
-      verb: Atom.to_string(route.verb)
+      verb: String.downcase(conn.method)
     }
   end
 
-  defp identify_route(conn) do
+  defp identify_route(%{method: method, host: host, path_info: path} = conn) do
     conn
-    |> router_module()
-    |> apply(:__routes__, [])
-    |> enum_find(conn)
+    |> router_module
+    |> apply(:__match_route__, [method, decode_uri(path), host])
+    |> extract_route_info()
   end
 
-  defp enum_find(routes, conn) do
-    routes
-    |> Enum.find(fn route -> has_eql_values?(route, conn) end)
-  end
+  defp router_module(%{private: %{phoenix_router: router}}), do: router
 
-  defp has_eql_values?(route, conn) do
-    route.plug == controller_module(conn) and router_options(route) == action_atom(conn) and
-      route.verb == verb_atom(conn) and match_path?(route, conn)
-  end
+  defp decode_uri(path_info), do: Enum.map(path_info, &URI.decode/1)
+
+  defp extract_route_info({%{} = route_info, _, _, _}), do: route_info
```

Comment:
WDYT in the `_` put the name of variable to make more understandable? even you are not using

Reply:
To be honest I don't know what this data is  =P  I used a metadata function from Phoenix that has this return:
```elixir
{%{
   conn: nil,
   log: :debug,
   path_params: %{},
   pipe_through: [:api],
   plug: Xcribe.UsersController,
   plug_opts: :index,
   route: "/users"
 }, #Function<56.91810571/2 in Xcribe.WebRouter.__match_route__/3>,
 #Function<57.91810571/1 in Xcribe.WebRouter.__match_route__/3>,
 {Xcribe.UsersController, :index}}
```
I'm not sure about this functions ref but I believe they are callbacks. I renamed them as you suggest as `callback_one` and `callback_tow`